Update AB and BSE and display counts for degree progress

bnehoran opened the pull request:
This pull request does the following two things:

1. Clean up the AB and BSE requirement files:
    - Remove the deprecated `description` field on subrequirements
    - Make the explanations more informative
    - Add a description for the overall degree (displayed in the req popup for the degree)
    - Fix a bug where ISC courses could only count for one requirement at a time
    - Add 1027 level to language requirement (for FRE 1027 and ITA 1027, double courses which complete the language to the 107 level)

2. Change the logic for degree progress requirements (those that just count the number of courses taken by a certain semester) so that they display the number of courses needed and the number taken, which is much more informative than the existing binary indicators.
This closes #166.
Before:
![image](https://user-images.githubusercontent.com/9757500/87890345-1dfcad00-ca04-11ea-89b1-be2cb4a8ea36.png)
After:
![image](https://user-images.githubusercontent.com/9757500/87870099-fd384700-c972-11ea-926f-05aa4b94bff7.png)

bnehoran committed:
Make degree progress show number of courses

Closes #166

bnehoran committed:
Clean up AB req file

bnehoran committed:
Clean up BSE req file

bnehoran committed:
Update AB degree progress requirements

bnehoran committed:
Update BSE degree progress requirements

bnehoran committed:
Add description text to the degrees

churichard commented on tigerpath/majors_and_certificates/degrees/AB_2018.json:
```diff
@@ -3,24 +3,50 @@
+          "max_counted": 1,
+          "explanation": null,
+          "completed_by_semester": 6,
+          "num_courses": 25
```
How'd you get 25 here?

bnehoran replied:
It's based on 
https://ua.princeton.edu/contents/academic-regulations/academic-year#:~:text=A.B.%20Program

But then again, here
https://ua.princeton.edu/contents/academic-standing-leaves-of-absence-and-reinstatement#:~:text=A.B.%20Program
it says 24 is the minimum.

It looks to me like 24 is a bare minimum, without which you get kicked out, while 25 is the expected number.

In the same way, you're expected to complete [8 and 17 courses](https://ua.princeton.edu/contents/academic-regulations/academic-year#:~:text=A.B.%20Program) by the end of the first two years respectively, but you only get kicked out if you go below [7 and 16](https://ua.princeton.edu/contents/academic-standing-leaves-of-absence-and-reinstatement#:~:text=A.B.%20Program).

churichard commented on tigerpath/majors_and_certificates/degrees/AB_2018.json:
```diff
@@ -31,8 +57,7 @@
       "min_needed": 1,
-      "description": "Writing Seminar — one course",
-      "explanation": "All Princeton undergraduates take a Writing Seminar in the Fall or Spring of the freshman year.\n\nThe Writing Seminars give Princeton freshmen an early opportunity to belong to a lively academic community in which members investigate a shared topic and discuss their writing together, with the aim of clarifying and deepening their thinking. The seminars are interdisciplinary ...
+      "explanation": "<b>Writing Seminar</b><br><i>one course, completed in the first year</i><hr>All Princeton undergraduates take a Writing Seminar in the Fall or Spring of the freshman year.\n\nThe Writing Seminars give Princeton freshmen an early opportunity to belong to a lively academic community in which members investigate a shared topic and discuss their writing together, with the aim of ...
```
That is some ghastly looking code...

bnehoran replied:
What do you suggest?

Does it at least look okay on the users end?

churichard replied:
Chatted with Barak - we decided to leave it for now

churichard commented on tigerpath/majors_and_certificates/degrees/BSE_2018.json:
```diff
@@ -3,53 +3,50 @@
+          "max_counted": 1,
+          "explanation": null,
+          "completed_by_semester": 2,
+          "num_courses": 9
```
How'd you get 9 here?

bnehoran replied:
I remember seeing somewhere that it said BSE students normally enroll in 4 and 5 courses in their first two semesters.
I don't remember where I saw this, and I can't find it again.

Would you prefer to change it to 8, which is the minimum according to these sources?
https://ua.princeton.edu/contents/academic-standing-leaves-of-absence-and-reinstatement#:~:text=B.S.E.%20Program
https://ua.princeton.edu/contents/program-study-degree-bachelor-science-engineering#:~:text=B.S.E.%20students%20enroll
and
https://ua.princeton.edu/contents/academic-regulations/academic-year#:~:text=B.S.E.%20Program

bnehoran replied:
```suggestion
          "num_courses": 8
```

churichard commented on tigerpath/majors_and_certificates/degrees/BSE_2018.json:
```diff
@@ -256,40 +241,36 @@
               "min_needed": null,
-              "description": "Epistemology and Cognition (EC)",
-              "explanation": "The requirement in epistemology and cognition will introduce you to the critical study of the nature, sources and bounds of human knowledge. Courses in this category examine topics ranging from the basic perceptual capabilities that humans share with other animals to the distinctively human capacity for language; from theoretical models of human knowledge to empirica ...
+              "explanation": "<b>Epistemology and Cognition (EC)</b><hr>The requirement in epistemology and cognition will introduce you to the critical study of the nature, sources and bounds of human knowledge. Courses in this category examine topics ranging from the basic perceptual capabilities that humans share with other animals to the distinctively human capacity for language; from theoret ...
```
Do we really need to repeat the title in the explanation? It's redundant, since the title is already in the popover header.

![Untitled](https://user-images.githubusercontent.com/4218237/87891302-598c7c80-c9ee-11ea-8854-a4f26d2eb30a.png)

bnehoran replied:
I honestly just wanted to include the two-letter code (EC), but it would look weird to just have that by itself.
Do you have an idea for how to do that?

churichard replied:
Chatted with Barak - we decided to include it in the explanation

bnehoran commented on tigerpath/majors_and_certificates/scripts/verifier.py:
```diff
@@ -336,8 +336,8 @@ def _init_req_fields(req):
-    elif "num_courses" in req and req["name"] == None and "completed_by_semester" in req:
-        req["name"] = "Complete " + str(req["num_courses"]) + " courses by Semester " + str(req["completed_by_semester"])
+    elif "num_courses" in req:
+        req["min_needed"] = req["num_courses"]
```
This, combined with line 538, causes it to report back the number of courses completed out of the total rather than just a yes/no.

bnehoran commented on tigerpath/majors_and_certificates/degrees/BSE_2018.json:
```diff
@@ -328,33 +306,30 @@
-              "explanation": "When you become proficient in a foreign language, you acquire more than a communication skill; you become literate in another culture and gain another perspective on the world.",
+              "explanation": null,
               "course_list": [
+                "LANG 1027",
```
French and Italian both have a 1027-level double course, which combines 102 and 107 and completes the language requirement, but was previously not counting.
This fixes it.

bnehoran commented on tigerpath/majors_and_certificates/degrees/BSE_2018.json:
```diff
@@ -328,33 +306,30 @@
               "min_needed": null,
-              "description": "Epistemology and Cognition (EC)",
-              "explanation": "The requirement in epistemology and cognition will introduce you to the critical study of the nature, sources and bounds of human knowledge. Courses in this category examine topics ranging from the basic perceptual capabilities that humans share with other animals to the distinctively human capacity for language; from theoretical models of human knowledge to empirica ...
+              "explanation": null,
```
No need for an explanation on these, since they are all hidden.

bnehoran commented on tigerpath/majors_and_certificates/degrees/BSE_2018.json:
```diff
@@ -171,16 +161,15 @@
-      "explanation": null,
+      "explanation": "<b>Chemistry</b><br><i>one semester, must be completed by the end of the first year</i><hr>207 Advanced General Chemistry: Materials Chemistry, or 201 General Chemistry, or equivalent",
       "pdfs_allowed": false,
+      "double_counting_allowed": true,
```
This and the `double_counting_allowed` on lines 119 and 201 fixes the problem where the ISC curriculum wasn't able to satisfy more than one requirement at a time.

Rather than just putting it only on the ISC subrequirement, I put it here, because otherwise, it conflicts with it being a hidden subrequirement (see https://github.com/TigerPathApp/tigerpath/issues/149). It doesn't matter because these other courses couldn't double count anywhere else anyway, so it only makes a difference for the ISC courses.

bnehoran commented on tigerpath/majors_and_certificates/degrees/BSE_2018.json:
```diff
@@ -60,16 +57,15 @@
-      "explanation": null,
+      "explanation": "<b>Mathematics</b><br><i>four semesters</i>",
       "pdfs_allowed": false,
+      "double_counting_allowed": true,
```
This is here just for consistency with the other prerequisites. It doesn't make a difference here either way.

bnehoran commented on tigerpath/majors_and_certificates/degrees/AB_2018.json:
```diff
@@ -59,71 +84,62 @@
-          "min_needed": null,
-          "description": "Science and Technology with Lab (STL)",
-          "explanation": null,
+          "min_needed": 1,
```
This turns it red, making clear that at least one such course is required. Has no other practical effect, since this is already taken care of by the `max_counted` on the STN subreq.

bnehoran commented on tigerpath/majors_and_certificates/degrees/AB_2018.json:
```diff
@@ -3,24 +3,50 @@
+        {
+          "name": "Total courses",
           "max_counted": 1,
-          "min_needed": 1,
```
`min_needed` is now ignored for degree progress requirements (see the edits to the verifier code below).
Instead, it is automatically set to whatever value `num_courses` has.

So I'm removing it here and in the BSE file as well.

bnehoran commented on tigerpath/majors_and_certificates/degrees/BSE_2018.json:
```diff
@@ -3,53 +3,50 @@
-          "name": null,
+          "name": "By first semester",
           "max_counted": 1,
-          "min_needed": 1,
```
This is now ignored. See the [comment on line 22 of the A.B. file](https://github.com/TigerPathApp/tigerpath/pull/355/files#r457024757).

bnehoran committed:
Update BSE first year num_courses to 8

bnehoran committed:
Remove redundant titles in the explanation

This commit removes the redundant titles in the explanation because 
those titles that were also in the explanation were redundant.

bnehoran committed:
Clean up BSE math explanations
